### ISSUE TITLE
Crate Geom - Intersections method not returning intersections on end points

### ISSUE THREAD

**Issue by Dollab** [nical/lyon#675]
Hi, the intersection algorithms in the **geom** crate seem to deliberately not return the intersections on the extremities of a path (eg. when the t parameter is equal to 1.0 or to 0.0). Is there a specific reason for this ? 

This makes it difficult to re-use the intersection algorithms for other purposes, as the expected behavior would be to return all intersections including those on the endpoints.

Would you be opened in changing this behavior in the library ? I would be happy to provide the PR.

For example, the following test will not return the two intersection points : 
```rust
#[test]
fn test_lyon_line_cubic_intersection_on_end_points() {
      // vertical line x = 0.0
      let line_segment = LineSegment {
          from: Point::new(0.0, -100.0),
          to: Point::new(0.0, 100.0),
      };

      // cubic with end points on x = 0.0
      let cubic = CubicBezierSegment {
          from: Point::new(0.0, 0.0),
          ctrl1: Point::new(20.0, 20.0),
          ctrl2: Point::new(20.0, 40.0),
          to: Point::new(0.0, 60.0),
      };

      let intersection = cubic.line_segment_intersections_t(&line_segment);
      
      assert!(intersection.len(), 2, "Two intersections point should be found");
      // assert_eq!(intersection[0], (0.0, 0.5));
      // assert_eq!(intersection[1], (1.0, 0.8));
}
```

**Comment by nical** [comment]
the intersection checks are used to detect path self-intersections in various algorithms of this repository where we don't want consecutive edges or other configurations where edges share a vertex) to count as intersecting. That's why the intersection checks only consider the curves without their exact endpoints.

For the most part I stand by this choice but it might make sense to only discard intersections that are exactly on endpoints of both curves (it would find the intersection in the middle of the edge in your example).
whenever I get some time I'll try and see how it affects the tessellation algorithm.

**Comment by nical** [comment]
In fact the behavior I described is already what's happening with line segment intersections (which is used in the trickiest parts of this repository). So I'll go ahead and align the other intersection functions with that.

I'm a hesitant to do it without an API bump since it's a fairly significant behavior change. That might delay things a bit.

**Comment by Dollab** [comment]
Thanks a lot, that would be great !
Lyon geom is one of the best crate I've found to deal with polybezier curves and I'm using it to implement some common algorithms needed for a drawing tool : nearest point, trim, extend, assemblies, offset, path boolean ops etc.
Would you be interested in extending this crate into a more complete geometric kernel with such functionalities ?

I was hoping such a crate already existed but to my knowledge it seems like it does not ?

**Comment by nical** [comment]
That is very nice to hear, thank you very much!

I'm happy to add this type of features to lyon. The geom crate is focused on curve segments, but there is a whole lot of these algorithms that would fit nicely in lyon_path or lyon_algorithms.

I don't know of crates providing a comprehensive set of path manipulation algorithms for drawing tools, but it is definitely stuff I am interested in.

**Comment by nical** [comment]
Fixed in #676 and published in lyon_geom 0.17.3.
